documentation/resolver_config.md
# resolver_config

[Back to resource list](../README.md#resources)

Introduced: v3.0.0

## Actions

- `:set`

## Properties

| Name                   | Type          | Default                          | Description                                                                     |
| ---------------------- | ------------- | -------------------------------- | --------------------------------------------------------------------------------|
| `config_file`          | String        | `/etc/resolv.conf`               | The path to the resolver configuration file on disk                             |
| `cookbook`             | String        | `resolver`                       | Cookbook to source configuration file template from                             |
| `template`             | String        | `resolv.conf.erb`                | Template to use to generate the configuration file                              |
| `owner`                | String        | Platform dependant               | Owner of the generated configuration file                                       |
| `group`                | String        | Platform dependant               | Group of the generated configuration file                                       |
| `mode`                 | String        | `'0644'`                         | Filemode of the generated configuration file                                    |
| `nameservers`          | String, Array | `nil`                            | The DNS servers to configure for system name resolution                         |
| `domain`               | String        | `nil`                            | The DNS domain name for the host                                                |
| `search`               | String, Array | `nil`                            | The DNS domain search list for the host                                         |
| `sortlist`             | String, Array | `nil`                            | The DNS sort list for the host                                                  |
| `options`              | Hash          | `nil`                            | Additional options to add to the resolver configuration file                    |
| `atomic_update`        | true, false   | `true`                           | Allow to Ran atomic file update (Could be used when updating inside a container |

## Examples

```ruby
resolver_config '/etc/resolv.conf' do
  nameservers ['1.1.1.1', '1.0.0.1']
  domain 'test.com'
  search ['test1.com', 'test2.com']
  options(
    'timeout' => 2
  )
end
```


documentation/resolver_systemd_resolved_config.md
# resolver_systemd_resolved_config

[Back to resource list](../README.md#resources)

Create and manage the systemd-resolved configuration file. (<https://www.freedesktop.org/software/systemd/man/resolved.conf.html>)

Introduced: v3.0.0

## Actions

- `:set`

## Properties

| Name                   | Type          | Default                          | Description                                                         |
| ---------------------- | ------------- | -------------------------------- | ------------------------------------------------------------------- |
| `config_file`          | String        | `/etc/systemd/resolved.conf`     | The path to the systemd-resolved configuration file on disk         |
| `cookbook`             | String        | `resolver`                       | Cookbook to source configuration file template from                 |
| `template`             | String        | `resolved.conf.erb`              | Template to use to generate the configuration file                  |
| `owner`                | String        | Platform dependant               | Owner of the generated configuration file                           |
| `group`                | String        | Platform dependant               | Group of the generated configuration file                           |
| `mode`                 | String        | `'0644'`                         | Filemode of the generated configuration file                        |
| `dns`                  | String, Array | `nil`                            | A list of IPv4 and IPv6 addresses to use as system DNS servers      |
| `fallback_dns`         | String, Array | `nil`                            | A list of IPv4 and IPv6 addresses to use as fallback DNS servers    |
| `domains`              | String, Array | `nil`                            | A list of DNS search domains for the host                           |
| `llmnr`                | True, False, String | `nil`                      | Control Link-Local Multicast Name Resolution support (RFC 4795) on the host |
| `multicast_dns`        | True, False, String | `'resolve'`                | Control Multicast DNS support (RFC 6762) on the host                |
| `dnssec`               | True, False, String | `'allow-downgrade'`        | Control DNSSEC validation on the host                               |
| `dns_over_tls`         | True, False, String | `'opportunistic'`          | Control DNS-over-TLS on the host                                    |
| `cache`                | True, False, String | `true`                     | Control DNS caching on the host                                     |
| `cache_from_localhost` | True, False   | `false`                          | Control DNS caching of a result from a local address on the host    |
| `dns_stub_listener`    | True, False, String | `nil`                      | Control the DNS stub listener on the host                           |
| `dns_stub_listener_extra` | True, False, String | `nil`                   |Additional DNS stub sockets to listen on for the host                |
| `read_etc_hosts`       | True, False   | `true`                           | Control resolution of a hostname from the /etc/hosts file           |
| `resolve_unicast_single_label` | True, False | `false`                    | Control resolution of A and AAAA queries for single-label names over classic DNS |

## Examples

```ruby
resolver_systemd_resolved_config '/etc/systemd/resolved.conf' do
  dns ['1.1.1.1', '1.0.0.1', '2606:4700:4700::1111', '2606:4700:4700::1001']
  fallback_dns '8.8.8.8,8.8.4.4,2001:4860:4860::8888,2001:4860:4860::8844'
  domains 'localtest.com'
end
```


CHANGELOG.md
# resolver Cookbook CHANGELOG

This file is used to list changes made in each version of the resolver cookbook.

## Unreleased

## 4.1.0 - *2022-08-08*

- Remove delivery and move to calling RSpec directly via a reusable workflow
- Add the `atomic_update` property
- Use MacOS 10.15 when using Vagrant testing
- Update Linux platforms for testing

## 4.0.3 - *2022-02-08*

- Remove delivery folder

## 4.0.2 - *2021-08-30*

- Standardise files with files in sous-chefs/repo-management

## 4.0.1 - *2021-06-01*

- Standardise files with files in sous-chefs/repo-management

## 4.0.0 - *2021-05-26*

- Switch to `unified_mode`
- Update `chef_version` to only support Chef 15.3 or later

## 3.0.4 - *2021-03-03*

- Fixed typo in the `CacheFromLocalhost` parameter in the systemd-resolved template

## 3.0.3 - *2021-02-25*

- Fix blog link for wrapper cookbooks

## 3.0.2 - *2021-02-25*

- Merge resolv.conf options & remove extra colons

## 3.0.1 - *2021-02-20*

- Update `Chef` -> `Chef Infra`

## 3.0.0 - *2021-02-18*

- Port recipes and attributes to resources
- Add support for overriding system configuration methods to resolver_config
- Add systemd-resolved support

## 2.2.0 - *2021-02-10*

- Sous Chefs Adoption
- Add InSpec Tests
- Remove Deprecated Platforms, Centos 6, Ubuntu 16.04, opensuse-42, fedora-29

## 2.1.0 (2018-07-18)

- Test with Local Delivery instead of Rake
- Update Apache license string in metadata
- the search stanza should be an array rather than a single element.  resolv.conf search allows up to six search domains to be defined.  Have updated the recipes with some rules (no more than 6 domains or 256 chars), change the ['resolver']['search'] into an array and changed teh template unless block to search for an empty array instead
- add aix support

## 2.0.1 (2017-02-24)

- Fix compatibility with old chef 12 clients
- Remove old recipe from the metadata
- Call out SUSE support in the readme and metadata

## 2.0.0 (2016-09-07)

- add support for 'domain' resolv.conf key
- Switch to root group to support AIX/OS X/FreeBSD
- Removed the from_server_role recipe. This should be done with a wrapper cookbook instead
- Add testing with Travis CI and use kitchen-dokken for integration testing
- Add basic Chefspec testing
- Add integration testing with Test Kitchen
- Add Berksfile
- Update contributing, testing and maintainers docs
- Improve the readme with chef/platform Requirements
- Add a Rakefile for simplified testing
- Use strings for file modes to preserve the leading 0s
- Add source_url, issue_url and chef_version to the metadata
- Remove attributes from the metadata since they weren't up to date

## v1.3.0 (2015-10-16)

- Add rubocop file and resolve all rubocop warnings
- Update the contributing doc to match the new process
- Use strings for file modes to preserve the leading 0s
- Add support the domain resolv.conf config option
- Add a gitignore and chefignore file
- Added a header to the resolv.conf file stating that chef edited the file
- Use the Ohai 'root_group' attribute for the config file to support non-Linux *nix platforms. This raises the minimum required version of Chef to 11.14.0 with Ohai 7.2.0.

## v1.2.0 (2014-10-16)

- [#11] Improve safety and sanity of cookbook
- [#12] README update
- [#14] Exclude search in template if attribute is empty

## v1.1.4

- **[COOK-3375](https://tickets.chef.io/browse/COOK-3375)** - Add Solaris support

## v1.1.2

### Bug

- [COOK-2977]: resolver cookbook has foodcritic failures

## v1.1.0

- [COOK-1242] - collect nameservers using a server role, adds `from_server_role` recipe

## v1.0.2

- 'node' went missing in the template.

## v1.0.0

- [COOK-1089] - accept resolver options via attributes
- [COOK-1109] - return from resolver recipe if no nameserver attribute is set so resolv.conf doesn't get broken during the run
- [COOK-1150] - drop the empty element from the nameserver attribute array


resources/config.rb
#
# Cookbook:: resolver
# Resource:: config
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

unified_mode true

include Resolver::Cookbook::Helpers

property :config_file, String,
          default: lazy { RESOLVER_RESOLV_CONF_DEFAULT },
          description: 'The path to the resolver configuration file on disk'

property :cookbook, String,
          default: 'resolver',
          description: 'Template source cookbook'

property :template, String,
          default: 'resolv.conf.erb',
          description: 'Template source file'

property :owner, String,
          default: lazy { RESOLVER_CONF_USER_DEFAULT },
          description: 'Generated file owner'

property :group, String,
          default: lazy { resolver_conf_group_default },
          description: 'Generated file group'

property :mode, String,
          default: '0644',
          description: 'Generated file mode'

property :nameservers, [String, Array],
          coerce: proc { |p| p.is_a?(String) ? p.split(',') : p },
          required: true,
          description: 'The DNS servers to configure for system name resolution'

property :domain, String,
          description: 'The DNS domain name for the host'

property :search, [String, Array],
          coerce: proc { |p| p.is_a?(String) ? p.split(',') : p },
          description: 'The DNS domain search list for the host',
          callbacks: {
            'with glibc < 2.26 the number of search domains are limited to 6' => lambda { |p|
              p.count <= 6 || (node['languages']['c']['glibc']['version'].to_f >= 2.26)
            },
            'with glibc < 2.26 the total length of the search list is limited to 256 character' => lambda { |p|
              p.join(' ').length <= 256 || (node['languages']['c']['glibc']['version'].to_f >= 2.26)
            },
          }

property :sortlist, [String, Array],
          coerce: proc { |p| p.is_a?(Array) ? p : [p] },
          description: 'The DNS sort list for the host'

property :options, Hash,
          description: 'Additional options to add to the resolver configuration file'

property :atomic_update, [true, false],
          default: true,
          description: 'Perform atomic file updates on a per-resource basis'

property :override_system_configuration, [true, false],
          default: lazy { resolver_override_default(config_file) },
          description: 'Override the system DNS configuration, for use with NetworkManager/resolvconf/systemd-resolved'

action :set do
  if new_resource.override_system_configuration
    directory '/etc/NetworkManager/conf.d' do
      recursive true
    end

    template '/etc/NetworkManager/conf.d/90-dns-none.conf' do
      cookbook new_resource.cookbook
      source '90-dns-none.conf.erb'

      owner new_resource.owner
      group new_resource.group
      mode new_resource.mode

      action :create
    end if ::Dir.exist?('/etc/NetworkManager')
  end

  template new_resource.config_file do
    cookbook new_resource.cookbook
    source new_resource.template

    owner new_resource.owner
    group new_resource.group
    mode new_resource.mode

    force_unlink new_resource.override_system_configuration
    manage_symlink_source !new_resource.override_system_configuration

    atomic_update new_resource.atomic_update

    variables(
      nameservers: new_resource.nameservers,
      domain: new_resource.domain,
      search: new_resource.search,
      sortlist: new_resource.sortlist,
      options: new_resource.options
    )
    helpers(Resolver::Cookbook::TemplateHelpers)

    action :create
  end
end
